src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -91,19 +100,72 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
             return null;
         }
 
+        // Because SemanticToken data is generated relative to the previous token it must be in order.
+        // We have a guarantee of order within any given language server, but when we translate from the ranges
+        // in the C# document to ranges into the Razor document we lose that guarantee.
+        // Having converted them to SemanticRange objects, we can simply do the final round of a merge sort.
         var newList = new List<SemanticRange>(ranges1.Count + ranges2.Count);
-        newList.AddRange(ranges1);
-        newList.AddRange(ranges2);
+        var idxRange1 = 0;
+        var idxRange2 = 0;
+        while (idxRange1 < ranges1.Count && idxRange2 < ranges2.Count)
+        {
+            var range1 = ranges1[idxRange1];
```

Comment:
I think it would be good to be really clear about the naming here, so that the algorithm is easier to understand, and rename `ranges1` to `razorRanges` and `ranges2` to `csharpRanges`